File: CYJmission04.py
import RPi.GPIO as GPIO
import time

# 사용할 GPIO 핀의 번호를 설정
button_pin1 = 22
button_pin2 = 23
button_pin3 = 24

# Define the pin numbers as global variables
STEP_OUTA = 16
STEP_OUTB = 17
STEP_OUT2A = 18
STEP_OUT2B = 19

GPIO.setwarnings(False)
GPIO.setmode(GPIO.BCM)  # 핀모드 설정

# 버튼 핀의 입력설정
GPIO.setup(button_pin1, GPIO.IN)
GPIO.setup(button_pin2, GPIO.IN)
GPIO.setup(button_pin3, GPIO.IN)

# Motor control pin numbers as outputs
GPIO.setup(STEP_OUTA, GPIO.OUT)
GPIO.setup(STEP_OUTB, GPIO.OUT)
GPIO.setup(STEP_OUT2A, GPIO.OUT)
GPIO.setup(STEP_OUT2B, GPIO.OUT)

def rotate_motor(steps, delay_time, step_outa, step_outb, step_out2a, step_out2b):
    for i in range(steps):
        GPIO.output(step_outa, 1)
        time.sleep(delay_time)
        GPIO.output(step_outa, 0)
        GPIO.output(step_outb, 1)
        time.sleep(delay_time)
        GPIO.output(step_outb, 0)
        GPIO.output(step_out2a, 1)
        time.sleep(delay_time)
        GPIO.output(step_out2a, 0)
        GPIO.output(step_out2b, 1)
        time.sleep(delay_time)
        GPIO.output(step_out2b, 0)

while True:
    if GPIO.input(button_pin1) == GPIO.LOW:
        rotate_motor(60, 0.01, STEP_OUTA, STEP_OUTB, STEP_OUT2A, STEP_OUT2B)  # Rotate 45 degrees
    if GPIO.input(button_pin2) == GPIO.LOW:
        rotate_motor(120, 0.01, STEP_OUTA, STEP_OUTB, STEP_OUT2A, STEP_OUT2B)  # Rotate 90 degrees
    if GPIO.input(button_pin3) == GPIO.LOW:
        rotate_motor(240, 0.01, STEP_OUTA, STEP_OUTB, STEP_OUT2A, STEP_OUT2B)  # Rotate 180 degrees

# Clean up the GPIO pins
GPIO.cleanup()

File: CYJmission04.c
#include <wiringPi.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define STEP_OUTA 27
#define STEP_OUTB 0
#define STEP_OUT2A 1
#define STEP_OUT2B 24
#define SW1 3
#define SW2 4
#define SW3 5

void rotateMotor(int steps, int delayTime) {
    int i;

    for (i = 0; i < steps; i++) {
        digitalWrite(STEP_OUTA, 1);
        delayMicroseconds(delayTime);
        digitalWrite(STEP_OUTA, 0);
        digitalWrite(STEP_OUTB, 1);
        delayMicroseconds(delayTime);
        digitalWrite(STEP_OUTB, 0);
        digitalWrite(STEP_OUT2A, 1);
        delayMicroseconds(delayTime);
        digitalWrite(STEP_OUT2A, 0);
        digitalWrite(STEP_OUT2B, 1);
        delayMicroseconds(delayTime);
        digitalWrite(STEP_OUT2B, 0);
    }
}

int main(void) {
    if (wiringPiSetup() == -1) {
        return 1;
    }

    pinMode(STEP_OUTA, OUTPUT);
    pinMode(STEP_OUTB, OUTPUT);
    pinMode(STEP_OUT2A, OUTPUT);
    pinMode(STEP_OUT2B, OUTPUT);
    pinMode(SW1, INPUT);
    pinMode(SW2, INPUT);
    pinMode(SW3, INPUT);

    while (1) {
        if (digitalRead(SW1) == LOW) {
            printf("SW1 Button pushed! Rotating motor 45 degrees.\n");
            rotateMotor(45*1.5, 2000);
        } else if (digitalRead(SW2) == LOW) {
            printf("SW2 Button pushed! Rotating motor 90 degrees.\n");
            rotateMotor(90*1.5, 2000);  
        } else if (digitalRead(SW3) == LOW) {
            printf("SW3 Button pushed! Rotating motor 180 degrees.\n");
            rotateMotor(180*1.5, 2000);  // 180도 회전
        }

        // 스위치 상태 체크 간격을 늘리기 위한 딜레이
        usleep(100000);  // 0.1초
    }

    return 0;
}

